### ISSUE TITLE
Add streaming search support

### ISSUE THREAD

**Issue by assafmo** [assafmo/xioc#8]
@philpraxis:
> On input, if non-allocating aka streaming search support could be implemented (it looks like not it's not, I didn't look at the source), that would be a huge plus since you _will_ have arbitrarily long input lines in your input.

**Comment by assafmo** [comment]
These might help:
- https://www.reddit.com/r/golang/comments/9htl34/streaming_regex_with_ioreader/
- https://golang.org/pkg/io/#MultiWriter

**Comment by philpraxis** [comment]
Hey Assaf, yes this looks good.

Indeed what looks "reasonable" in one context in term of baseline allocation often becomes limited in another context.

Best

**Comment by assafmo** [comment]
**TL;DR:**  
I think I'll keep the line reader for now, with a 5MiB limit per line.  
I'll also try to gracefully handle very long lines, without crashing. :laughing: 

<hr>

@philpraxis So I've implemented streaming search just like you suggested.  
It's about two times slower than the current version:
```bash
➜  xioc git:(master) ✗ pv test.txt | xioc -o ip4 > line-reader.txt
 113MiB 0:00:24 [4.57MiB/s] [====================================================>] 100%
➜  xioc git:(master) ✗ pv test.txt | go run main.go -o ip4 > byte-reader.txt
 113MiB 0:00:40 [2.84MiB/s] [====================================================>] 100%
```
I think it's because the way go exposes [regexp.FindReaderIndex](https://golang.org/pkg/regexp/#Regexp.FindReaderIndex). 
When using this function I've used channels and goroutines to create a friendly API like so:  
```go
for ip := range xioc.ExtractReaderIPv4s(os.Stdin){
    // ...
}
```
Instead of something like:
```go
scanner := xioc.ExtractReaderIPv4s(os.Stdin)
for scanner.Next() {
    ip := scanner.Text()
    // ...
}
```
So maybe this causes the slowness. See the [streaming_search](https://github.com/assafmo/xioc/tree/streaming_search) branch if you are interested.

**Comment by assafmo** [comment]
@philpraxis, commit 7025ded7546491e10e7de0de078c5dcee89e0647 fixes the issue you described in #7.  
Now `xioc v1.1.11` never crashes on very long lines.

I think this is what you meant by "stream searching", so I've decided to close this issue (Also after seeing that "true" stream searching is much slower than the current solution).

I'd love to know from you if this fixes your use case for very long lines.

**Comment by philpraxis** [comment]
Works good!! Thanks a lot!
